### PR TITLE
Get rid of remainder of libsoup usage in flatpak-builder

### PR DESCRIPTION
**Pull request by nanonyme**
Fixes https://github.com/flatpak/flatpak-builder/issues/477

### CODE REVIEW

**Comment by nanonyme** [comment]
@alexlarsson @TingPing this is probably awful stylically and haven't yet run this through valgrind but before continuing need decision what to do with glib compat. I am still working on validating everything works. Does flatpak-builder need to keep supporting glib 2.44? If it can be bumped to 2.66 (involves bumping builders to 22.04), we don't need compat layer in flatpak-builder which keeps code simpler.

File: src/builder-context.h [hunk 1]
```diff
@@ -22,14 +22,16 @@
...
-#include <libsoup/soup.h>
 #include <curl/curl.h>
 #include "builder-options.h"
 #include "builder-utils.h"
 #include "builder-sdk-config.h"
 
 G_BEGIN_DECLS
 
+/* Same as SOUP_HTTP_URI_FLAGS, means all possible flags for http uris */
+#define CONTEXT_HTTP_URI_FLAGS (G_URI_FLAGS_HAS_PASSWORD | G_URI_FLAGS_ENCODED_PATH | G_URI_FLAGS_ENCODED_QUERY | G_URI_FLAGS_ENCODED_FRAGMENT | G_URI_FLAGS_SCHEME_NORMALIZE)
```
Comment (smcv) [inline src/builder-context.h:33]: `GUri` with `G_URI_FLAGS_SCHEME_NORMALIZE` requires bumping `GLIB_REQS` in `configure.ac` to 2.68 (see flatpak/flatpak#5062 and flatpak/flatpak#5064), which would rule out compiling upstream source on Debian 11.

Alternatively, if it's OK to use `G_URI_FLAGS_SCHEME_NORMALIZE` if and only if it exists (as in flatpak/flatpak#5064), then that use of `GUri` will depend on GLib 2.66, which is available in Debian 11.

**Review by smcv — CHANGES_REQUESTED** [review]

File: src/builder-context.h (hunk 1 above)
Comment (nanonyme) [reply src/builder-context.h:33]: Yeah, I know the failure reason. Trying to raise some discussion on it. It is clear flatpak needs to support ancient distros but with flatpak-builder situation has been more fuzzy.

Comment (nanonyme) [reply src/builder-context.h:33]: But yeah, maybe not raising bar above 2.66 would be a good call anyway. This build runs with Ubuntu 20.04 which I suppose has even older glib.

Comment (nanonyme) [reply src/builder-context.h:35]: Right. I now did this change, there's obvious error for logs `Requested 'glib-2.0 >= 2.66' but version of GLib is 2.64.6`

Comment (nanonyme) [reply src/builder-context.h:35]: Does flatpak require glib 2.66 to build?

Comment (smcv) [reply src/builder-context.h:35]: Flatpak itself has a backport of GUri for GLib older than 2.66. The problem in flatpak/flatpak#5062 is that if GLib is *exactly* 2.66, as in Debian 11, the backport isn't used but the real OS GLib doesn't support `G_URI_FLAGS_SCHEME_NORMALIZE`.

Comment (smcv) [reply src/builder-context.h:35]: > It is clear flatpak needs to support ancient distros but with flatpak-builder situation has been more fuzzy

Flatpak allegedly supports ancient distros, but it doesn't *actually* support (sufficiently) ancient distros: flatpak/flatpak#5066.

For flatpak-builder and other ancillary stuff, I personally think a reasonable place to draw the line is the latest Debian stable release or the latest Ubuntu LTS, whichever is older. They're both released on a cycle of about 2 years, with an offset of 1 year, so the one that is the limiting factor swaps periodically: it's currently Debian 11, but when Debian 12 is released, Ubuntu 22.04 will become the limiting factor.

For Flatpak itself, clearly there's interest in supporting older distros in order to be able to use Flatpak to run recent software on old distros, and in particular people do use it on older-than-latest-LTS branches of Ubuntu (although I'm the only one who is actually working on those backports, and I don't use Ubuntu myself). I think we should seriously consider dropping any attempt to backport to Ubuntu 16.04 (which is EOL for Ubuntu's standard support, and requires a support subscription for extended security maintenance), and perhaps also 18.04 (which is not EOL, but Flatpak doesn't compile successfully there and nobody noticed until I tried to update the backport).

Comment (nanonyme) [reply src/builder-context.h:35]: Seems there is no flatpak backports for jammy.

Comment (nanonyme) [reply src/builder-context.h:35]: I did not touch that part of build that did `sudo add-apt-repository ppa:alexlarsson/flatpak` yet.

Comment (nanonyme) [reply src/builder-context.h:35]: Appears build is also doing `sudo add-apt-repository 'deb https://download.mono-project.com/repo/ubuntu stable-jammy main' # Needed for updates to work`. I updated name, seems there is no package for jammy. I don't know what we need from there.

Comment (smcv) [reply src/builder-context.h:35]: Adding the Mono repository was a workaround for a bug in how Github set up their containers: see <https://github.com/flatpak/flatpak/pull/4299>. If we don't need to use that workaround any more in 22.04, then it's better to stop using it (remove that line). We only want that workaround if it's still necessary.

Comment (smcv) [reply src/builder-context.h:35]: It's probably actually better if we do each CI build with a minimum of backports, and only add extra non-Ubuntu repositories as we find that we need them. Ubuntu 22.04 has Flatpak 1.12.x, which seems like it should be sufficient for flatpak-builder for a while.

Comment (nanonyme) [reply src/builder-context.h:35]: Looks like Python 2 test is skipped completely currently since build machine does not have Python 2 installed.

Comment (nanonyme) [reply src/builder-context.h:35]: @smcv I thought I did the requested changes but GitHub is still complaining. Did I miss something?

File: .github/workflows/check.yml [hunk 2]
```diff
@@ -11,12 +11,10 @@ on:
 jobs:
   check:
     name: Build with gcc and test
-    runs-on: ubuntu-20.04
+    runs-on: ubuntu-22.04
```
Comment (TingPing) [inline .github/workflows/check.yml:14]: This can be a separate commit.

File: configure.ac [hunk 3]
```diff
@@ -12,7 +12,7 @@ AC_INIT([flatpak-builder],
         [flatpak-builder],
         [http://flatpak.org/])
 
-GLIB_REQS=2.44
+GLIB_REQS=2.66
```
Comment (TingPing) [inline configure.ac:15]: This is a massive bump. I don't think its acceptable.

Comment (smcv) [reply configure.ac:15]: What older OSs are you interested in running flatpak-builder on? When there's a GLib feature that we can benefit from as much as the existence of GUri, I think the required version should be driven by which OSs we want to support, rather than by backwards-compat for the sake of backwards-compat.

Comment (gasinvein) [reply configure.ac:15]: It's flatpak-builder, not flatpak. Do we really need to support _building_ flatpaks on something older than Debian Stable? There is `org.flatpak.Builder` flatpak after all.

File: .github/workflows/check.yml (hunk 2 above)
Comment (nanonyme) [reply .github/workflows/check.yml:14]: Okay, something to keep in mind these changes are interdependent. Glib requires builder bump. But I can split.

File: configure.ac (hunk 3 above)
Comment (TingPing) [reply configure.ac:15]: I didn't track down who has it. Debian Stable having it is a reasonable metric for this being OK.

File: .github/workflows/check.yml (hunk 2 above)
Comment (nanonyme) [reply .github/workflows/check.yml:14]: I have now split it so that CI is updated as first commit and this work is on top of that.

**Comment by nanonyme** [comment]
So are there still further changes requested?

File: src/builder-context.c [hunk 4]
```diff
@@ -383,24 +382,31 @@ builder_context_download_uri (BuilderContext *self,
...
-  g_autoptr(SoupURI) original_uri = soup_uri_new (url);
   g_autoptr(GError) first_error = NULL;
+  g_autoptr(GUri) original_uri = g_uri_parse (url, CONTEXT_HTTP_URI_FLAGS, &first_error);
 
-  if (original_uri == NULL)
-    return flatpak_fail (error, _("Could not parse URI “%s”"), url);
+  if (original_uri == NULL) {
+    g_propagate_error (error, g_steal_pointer (&first_error));
+    return FALSE;
+  }
```
Comment (smcv) [inline src/builder-context.c:391]:
```
  if (original_uri == NULL)
    {
      g_propagate_error (error, g_steal_pointer (&first_error));
      return FALSE;
    }
```

File: src/builder-context.h [hunk 5]
```diff
@@ -22,14 +22,22 @@
...
 #include <curl/curl.h>
 #include "builder-options.h"
 #include "builder-utils.h"
 #include "builder-sdk-config.h"
 
 G_BEGIN_DECLS
 
+/* Same as SOUP_HTTP_URI_FLAGS, means all possible flags for http uris */
+
+#if GLIB_CHECK_VERSION (2, 68, 0) || !GLIB_CHECK_VERSION (2, 66, 0)
```
Comment (smcv) [inline src/builder-context.h:34]:
```
#if GLIB_CHECK_VERSION (2, 68, 0)
```

Unlike Flatpak, flatpak-builder no longer supports versions older than 2.66 after this commit, and doesn't have a backport of GUri.

File: src/builder-main.c [hunk 6]
```diff
@@ -553,7 +545,7 @@ main (int    argc,
   if (opt_sources_urls)
     {
       g_autoptr(GPtrArray) sources_urls = NULL;
-      sources_urls = g_ptr_array_new_with_free_func ((GDestroyNotify)soup_uri_free);
+      sources_urls = g_ptr_array_new_with_free_func ((GDestroyNotify)g_free);
```
Comment (smcv) [inline src/builder-main.c:548]:
```
      sources_urls = g_ptr_array_new_with_free_func ((GDestroyNotify) g_uri_unref);
```

File: src/builder-main.c [hunk 7]
```diff
@@ -562,7 +554,7 @@ main (int    argc,
               opt_sources_urls[i] = g_strdup_printf ("%s/", tmp);
             }
 
-          SoupURI *uri = soup_uri_new (opt_sources_urls[i]);
+          GUri *uri = g_uri_parse (opt_sources_urls[i], CONTEXT_HTTP_URI_FLAGS, &error);
           if (uri == NULL)
             {
               g_printerr ("Invalid URL '%s'", opt_sources_urls[i]);
```
Comment (smcv) [inline src/builder-main.c:560]: You probably want to show `error->message` here?

File: src/builder-source-file.c [hunk 8]
```diff
@@ -329,23 +328,26 @@ download_data_uri (const char     *url,
...
+  g_autoptr(GError) first_error = NULL;  
   g_autoptr(GInputStream) input = NULL;
   g_autoptr(GOutputStream) out = NULL;
+  g_autoptr(GUri) parsed = NULL;
 
-  session = builder_context_get_soup_session (context);
+  parsed = g_uri_parse(url, CONTEXT_HTTP_URI_FLAGS, error);
+  if (!parsed) {
+    return FALSE;
+  }
```
Comment (smcv) [inline src/builder-source-file.c:340]:
```
  if (!parsed)
    return FALSE;
```

or

```
  if (!parsed)
    {
      return FALSE;
    }
```

File: src/builder-source-file.c [hunk 9]
```diff
@@ -329,23 +328,26 @@ download_data_uri (const char     *url,
...
+  session = builder_context_get_curl_session (context);
+  out = g_memory_output_stream_new_resizable ();
 
-  input = soup_request_send (req, NULL, error);
-  if (input == NULL)
+  if (!builder_download_uri_buffer(parsed, session, out, NULL, 0, error))
+    {
+    g_propagate_error (error, g_steal_pointer (&first_error));
     return NULL;
+    }
```
Comment (smcv) [inline src/builder-source-file.c:349]:
```
    {
      g_propagate_error (error, g_steal_pointer (&first_error));
      return NULL;
    }
```

File: src/builder-context.h [hunk 10]
```diff
@@ -22,14 +22,22 @@
...
+#define CONTEXT_HTTP_URI_FLAGS (G_URI_FLAGS_HAS_PASSWORD | G_URI_FLAGS_ENCODED_PATH | G_URI_FLAGS_ENCODED_QUERY | G_URI_FLAGS_ENCODED_FRAGMENT | G_URI_FLAGS_SCHEME_NORMALIZE)
+#else
+/* GLib 2.66 didn't support scheme-based normalization */
```
Comment (TingPing) [inline src/builder-context.h:37]: We don't actually care unless we are doing comparisons between URIs. (We may or may not, I haven't looked).

File: src/builder-source-file.c [hunk 11]
```diff
@@ -329,23 +328,26 @@ download_data_uri (const char     *url,
...
 
-  req = soup_session_request (session, url, error);
-  if (req == NULL)
-    return NULL;
+  session = builder_context_get_curl_session (context);
+  out = g_memory_output_stream_new_resizable ();
 
-  input = soup_request_send (req, NULL, error);
-  if (input == NULL)
+  if (!builder_download_uri_buffer(parsed, session, out, NULL, 0, error))
```
Comment (TingPing) [inline src/builder-source-file.c:345]:
```suggestion
  if (!builder_download_uri_buffer (parsed, session, out, NULL, 0, error))
```

File: src/builder-context.h (hunk 5 above)
Comment (TingPing) [inline src/builder-context.h:34]: I'm not saying we care but its worth noting CURL has a complete API to parse URIs and GUri does nothing unique for us.

But of course the CURL API is the CURL API so its a lot uglier than GUri. https://curl.se/libcurl/c/libcurl-url.html

Comment (nanonyme) [reply src/builder-context.h:34]: Considering large amounts of this code is almost copy-paste from flatpak porting away from libsoup2, I would prefer not going into a larger refactoring into CURL API which I don't know well.

File: src/builder-context.h (hunk 10 above)
Comment (nanonyme) [reply src/builder-context.h:37]: Sure, this was mostly copy-paste from flatpak which did similar migration earlier. Afaik the thought behind is that with at least newer versions of glib you get behaviour that is equivalent to the old behaviour.

**Review by TingPing — APPROVED** [review]
This seems good to me, is it ready to merge @nanonyme ?

**Comment by nanonyme** [comment]
Well, I was trying it out to build various apps, didn't notice anything weird. There's a risk I didn't have complex enough apps to test through all possible scenarios.
